### PR TITLE
unix: Add FreeDos build (DJGPP)

### PR DESCRIPTION
**Pull request by pohmelie**

### CODE REVIEW

**Comment by pfalcon** [comment]
@dpgeorge: So, the main question is whether it deserves its own top-level dir or not (it was my suggestion to @pohmelie to start with having it inside unix/).

@pohmelie : Thanks for the patch, it's clear that there's still some background work to do to minimize duplication it requires (I'll look into that as time permits).

**Comment by dpgeorge** [comment]
unix/ is getting pretty convoluted now with all the different configuration options.  Ok, this patch is very clean, but still it's going in the direction of making the Makefile even harder to understand.

As a related note, I really *don't* want to merge #1619 because it just makes unix/ even more spaghetti like.  So for that (a cross compiler) I'd rather a new top-level dir.

Back to this PR, I think it's clean enough to stay in unix/.

File: unix/Makefile [hunk 1]
```diff
@@ -108,6 +114,19 @@ SRC_MOD += modjni.c
...
+	main.c \
+	gccollect.c \
+	unix_mphal.c \
+	input.c \
+	file.c \
+	modos.c \
+	alloc.c \
+	coverage.c \
+	$(SRC_MOD)
+else
```
Comment (dpgeorge) [inline unix/Makefile]: Probably this can be simplified/deduplicated.

Comment (pfalcon) [inline unix/Makefile]: Yes, that's what I head in mind in comment to @pohmelie . if logic likely can be cleaned up too.

**Comment by pfalcon** [comment]
> Back to this PR, I think it's clean enough to stay in unix/.

The question is actually the opposite - can a new top-level dir be created for this port, similar to windows? That generally involves more duplication - e.g., duplicating Makefile, that's why I suggested to conservatively start with minimizing the duplication.

File: unix/Makefile (hunk 1 above)
Comment (pohmelie) [inline unix/Makefile]: I'm not familar with makefiles, so how can we «branch»  oneliner (AFAIK "\" is "concat with next line" like in python). Or you mean break hardcoded module to `SRC_MOD`?

Comment (pfalcon) [inline unix/Makefile]: If it's all about "no moduselect for freedos", then moduselect.c should be #if'ed for MICROPY_PY_USELECT (which needs to be added to mpconfigport*.h (not py/mpconfig.h)). That's the preference we have - doing stuff on C level instead of Makefile level if possible.

This should be a separate patch if you go for it.

Comment (dhylands) [inline unix/Makefile]: You should do SRC_C =  to specify common modules and then use SRC_C += inside an ifneq to freedos to add moduselect

Comment (pohmelie) [inline unix/Makefile]: @pfalcon, ok, I will try.

Comment (pohmelie) [inline unix/Makefile]: Hm, so what to do? @pfalcon said not makefile, but C level, and @dhylands said to make some makefile branching.

Comment (dpgeorge) [inline unix/Makefile]: All other configurable modules are configured in mpconfigport.mk.  Perhaps it's best to copy that pattern (see eg how modtime.c is done with MICROPY_PY_TIME).

Comment (pfalcon) [inline unix/Makefile]: @pohmelie : you asked how to do it on makefile level, @dhylands told you, that doesn't imply how it should be done in this case ;-)

 > All other configurable modules are configured in mpconfigport.mk.

Well, not quite. mpconfigport.mk configures what can't be configured via mpconfigport.h (and that's specifically stuff which requires linking with extra libraries). Well, there're some weird exceptions (which probably should be cleaned up).

**Comment by dpgeorge** [comment]
> The question is actually the opposite - can a new top-level dir be created for this port, similar to windows?

Is freedos popular?  Adding a new top-level dir should not be done lightly.  In my mind I'm thinking that for any future MCU ports we would create a new repository (eg micropython/baremetal) to put them, and make micropython/micropython a git submodule.  Even though having freedos within unix/ is a bit confusing (who would know to look there?) I think the fact that the changes are minimal (1 new file, small mods to Makefile) means it should go there.

**Comment by pohmelie** [comment]
It looks like there is no possibility to exclude moduselect.c via mpconfig*.h (except #if whole content inside moduselect.c). Cause if we don't include uselect in mpconfigport_freedos.h:
```
#if MICROPY_PY_USELECT
#define MICROPY_PY_USELECT_DEF { MP_ROM_QSTR(MP_QSTR_uselect), MP_ROM_PTR(&mp_module_uselect) },
#else
#define MICROPY_PY_USELECT_DEF
#endif

#define MICROPY_PORT_BUILTIN_MODULES \
    MICROPY_PY_FFI_DEF \
    MICROPY_PY_JNI_DEF \
    MICROPY_PY_TIME_DEF \
    MICROPY_PY_SOCKET_DEF \
    { MP_ROM_QSTR(MP_QSTR_uos), MP_ROM_PTR(&mp_module_os) }, \
    MICROPY_PY_USELECT_DEF \
    MICROPY_PY_TERMIOS_DEF \

...

```
It's trying to compile anyway, since moduselect.c still in makefile sources.

**Comment by dpgeorge** [comment]
> except #if whole content inside moduselect.c

This is what we usually do, see eg py/modarray.c.

**Comment by pohmelie** [comment]
Updated. Have no success with C-level branching.

**Comment by dhylands** [comment]
The part that you're probably missing is that most of the file has a #if SOME_CONFIG_OPTION/#endif // SOME_CONFIG_OPTION around it.

You need to #include "py/mpconfig.h" before the above #if.

Then if SOME_CONFIG_OPTION is turned off, you wind up comping what amounts to an empty source file.

**Comment by pohmelie** [comment]
@dhylands 
> You need to #include "py/mpconfig.h" before the above #if.

But py/mpconfig.h have no idea what moduselect.c is and have no switches for it. Are you sure this should work?

**Comment by dhylands** [comment]
You would add a config option like MICROPY_MODULE_USELECT and that would be the SOME_CONFIG_OPTION I was referring to.

**Comment by pohmelie** [comment]
@dhylands add where? moduselect is unix-only module. The only place is makefile level or I missed something?

**Comment by pfalcon** [comment]
>  add where?

That question was proactively answered in first review comments:

https://github.com/micropython/micropython/pull/1705#discussion_r47413905

Generally, feel free to leave moduselect thing to me.

**Comment by pohmelie** [comment]
@pfalcon, ok, I will wait for your activity. This is too hard for me :confused:

**Comment by pfalcon** [comment]
@pohmelie : Here's the easiest way to do it: 2d11b174192e77c46df2849647a3ae661d885852

And FYI, the easiest way to use that setting is `make CFLAGS_EXTRA="-DMICROPY_PY_USELECT=0" -B` - that will rebuild unix port without uselect module, without making any changes to uPy source. (Just FYI, so you understand better how uPy config system works.)

**Comment by pohmelie** [comment]
Updated.

File: unix/mpconfigport_freedos.h [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+void mp_unix_alloc_exec(mp_uint_t min_size, void** ptr, mp_uint_t *size);
+void mp_unix_free_exec(void *ptr, mp_uint_t size);
+void mp_unix_mark_exec(void);
+#define MP_PLAT_ALLOC_EXEC(min_size, ptr, size) mp_unix_alloc_exec(min_size, ptr, size)
+#define MP_PLAT_FREE_EXEC(ptr, size) mp_unix_free_exec(ptr, size)
+
+#include <unistd.h>
+#define MP_PLAT_PRINT_STRN(str, len) do { ssize_t ret = write(1, str, len); (void)ret; } while (0)
+
+#ifdef __linux__
```
Comment (dpgeorge) [inline unix/mpconfigport_freedos.h]: Does this make sense to keep?

File: unix/mpconfigport_freedos.h [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+// Can access physical memory using /dev/mem
+#define MICROPY_PLAT_DEV_MEM  (1)
+#endif
+
+// Assume that select() call, interrupted with a signal, and erroring
+// with EINTR, updates remaining timeout value.
+#define MICROPY_SELECT_REMAINING_TIME (1)
+
+#ifdef __ANDROID__
```
Comment (dpgeorge) [inline unix/mpconfigport_freedos.h]: Does this make sense to keep?  Can you have android on freedos?

File: unix/mpconfigport_freedos.h [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+#define MICROPY_PORT_ROOT_POINTERS \
+    const char *readline_hist[50]; \
+    mp_obj_t keyboard_interrupt_obj; \
+    void *mmap_region_head; \
+
+// We need to provide a declaration/definition of alloca()
+// unless support for it is disabled.
+#if !defined(MICROPY_NO_ALLOCA) || MICROPY_NO_ALLOCA == 0
+#ifdef __FreeBSD__
```
Comment (dpgeorge) [inline unix/mpconfigport_freedos.h]: Does this make sense to keep, freebsd+freedos?

File: unix/Makefile [hunk 5]
```diff
@@ -30,17 +30,23 @@ else
...
-# while cross-compile ports require gcc, so we test here for OSX and 
-# if necessary override the value of 'CC' set in py/mkenv.mk
-ifeq ($(UNAME_S),Darwin)
-CC = clang
-# Use clang syntax for map file
-LDFLAGS_ARCH = -Wl,-map,$@.map
+ifeq ($(TARGET), freedos)
+ CC = $(DJGPP)/bin/i586-pc-msdosdjgpp-gcc
+ CROSS_COMPILE = $(DJGPP)/i586-pc-msdosdjgpp/bin/
+ CFLAGS_MOD += -DMICROPY_NLR_SETJMP -Dtgamma=gamma -DMICROPY_EMIT_X86=0 -DMICROPY_NO_ALLOCA=1 -DMICROPY_PY_USELECT=0
```
Comment (dpgeorge) [inline unix/Makefile]: Instead of doing this, is it possible to pass CC=... and CROSS_COMPILE=... and CFLAGS_EXTRA=... in the recursive call to make below?

File: unix/Makefile [hunk 6]
```diff
@@ -169,6 +177,9 @@ fast:
 minimal:
 	$(MAKE) COPT="-Os -DNDEBUG" CFLAGS_EXTRA='-DMP_CONFIGFILE="<mpconfigport_minimal.h>"' BUILD=build-minimal PROG=micropython_minimal MICROPY_PY_TIME=0 MICROPY_PY_TERMIOS=0 MICROPY_PY_SOCKET=0 MICROPY_PY_FFI=0 MICROPY_USE_READLINE=0
 
+freedos:
+	$(MAKE) TARGET=freedos CFLAGS_EXTRA='-DMP_CONFIGFILE="<mpconfigport_freedos.h>"' BUILD=build-freedos PROG=micropython_freedos MICROPY_PY_SOCKET=0 MICROPY_PY_FFI=0 MICROPY_PY_JNI=0 MICROPY_PY_USELECT=0
```
Comment (dpgeorge) [inline unix/Makefile]: Following on from question above: can this be changed to something like:
```
$(MAKE) \
  CC="$(DJGPP)/bin/i586-pc-msdosdjgpp-gcc" \
  CROSS_COMPILE="$(DJGPP)/i586-pc-msdosdjgpp/bin/" CFLAGS_EXTRA='-DMP_CONFIGFILE="<mpconfigport_freedos.h>" -DMICROPY_NLR_SETJMP -Dtgamma=gamma -DMICROPY_EMIT_X86=0 -DMICROPY_NO_ALLOCA=1 -DMICROPY_PY_USELECT=0' \
  MICROPY_PY_SOCKET=0 MICROPY_PY_FFI=0 MICROPY_PY_JNI=0 \
  BUILD=build-freedos \
  PROG=micropython_freedos
```
?

File: unix/mpconfigport_freedos.h (hunk 2 above)
Comment (pfalcon) [inline unix/mpconfigport_freedos.h]: So, I decided to not propose mpconfigport_freedos.h to just include mpconfigport.h and then override needed settings. But then the best suggestion I have is to make mpconfigport_freedos.h to contain minimal changes wrt mpconfigport.h, so even if changes are not made parallel to both, then they can be propagated later in batch easily.

File: unix/Makefile (hunk 5 above)
Comment (pfalcon) [inline unix/Makefile]: I'd also say there should not be specific paths like $(DJGPP)/bin/i586-pc-msdosdjgpp-gcc , it should be just assumed that crosscompiler is in PATH. Counter-arguments are welcome.

File: unix/Makefile (hunk 6 above)
Comment (pohmelie) [inline unix/Makefile]: It looks like it does not work.
```
freedos:
	$(MAKE) \
	CC=i586-pc-msdosdjgpp-gcc
	CFLAGS_EXTRA='-DMP_CONFIGFILE="<mpconfigport_freedos.h>"' \
	CFLAGS_MOD='-DMICROPY_NLR_SETJMP -Dtgamma=gamma -DMICROPY_EMIT_X86=0 -DMICROPY_NO_ALLOCA=1 -DMICROPY_PY_USELECT=0' \
	BUILD=build-freedos \
	PROG=micropython_freedos \
	MICROPY_PY_SOCKET=0 \
	MICROPY_PY_FFI=0 \
	MICROPY_PY_JNI=0 \
	MICROPY_PY_USELECT=0
```
```
freedos:
	$(MAKE) \
	CC=i586-pc-msdosdjgpp-gcc
	CFLAGS_EXTRA='-DMP_CONFIGFILE="<mpconfigport_freedos.h>" -DMICROPY_NLR_SETJMP -Dtgamma=gamma -DMICROPY_EMIT_X86=0 -DMICROPY_NO_ALLOCA=1 -DMICROPY_PY_USELECT=0' \
	BUILD=build-freedos \
	PROG=micropython_freedos \
	MICROPY_PY_SOCKET=0 \
	MICROPY_PY_FFI=0 \
	MICROPY_PY_JNI=0 \
	MICROPY_PY_USELECT=0
```
Same error for both:
```
$ make freedos
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
make \
        CC=i586-pc-msdosdjgpp-gcc
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
make[1]: Entering directory `/home/broomrider/tmp/micropython/unix'
GEN build/genhdr/qstrdefs.generated.h
In file included from ../py/mpconfig.h:45:0,
                 from <stdin>:27:
./mpconfigport.h:250:20: fatal error: alloca.h: No such file or directory
compilation terminated.
ERROR: Empty preprocessor output - check for errors above
make[1]: *** [build/genhdr/qstrdefs.generated.h] Error 1
make[1]: *** Deleting file `build/genhdr/qstrdefs.generated.h'
make[1]: Leaving directory `/home/broomrider/tmp/micropython/unix'
make: *** [freedos] Error 2
```

Comment (dhylands) [inline unix/Makefile]: It would be useful to see the output of:
```
make freedos V=1
```

**Comment by pohmelie** [comment]
Found error, I missed backslash in make :grin: 
Updated, but now there is some error with uos module (after https://github.com/micropython/micropython/commit/d874702fe1e5822f8a3e1662657ac27d0b15bb68#diff-e91d2d7c65b84fc761769dbaae125ed4R187) cause djgpp `dirent` struct have no `d_ino` field (http://www.delorie.com/djgpp/doc/libc/libc_646.html)

**Comment by dpgeorge** [comment]
@pohmelie thanks, the PR now looks beautiful!  We'll sort out the uos problem.

**Comment by dpgeorge** [comment]
@pfalcon regarding lack of d_ino under freedos... do you think it reasonable to make our own macro feature test called `_DIRENT_HAVE_D_INO` and then add the patch:
```diff
--- a/unix/modos.c
+++ b/unix/modos.c
@@ -184,7 +184,11 @@ STATIC mp_obj_t listdir_next(mp_obj_t self_in) {
     // DT_UNKNOWN should have 0 value on any reasonable system
     t->items[1] = 0;
     #endif
+    #if !defined(_DIRENT_HAVE_D_INO) || _DIRENT_HAVE_D_INO
     t->items[2] = MP_OBJ_NEW_SMALL_INT(dirent->d_ino);
+    #else
+    t->items[1] = MP_OBJ_NEW_SMALL_INT(0);
+    #endif
     return MP_OBJ_FROM_PTR(t);
 }
```
?
(This means you have to define `_DIRENT_HAVE_D_INO` to 0 to force it off.)

**Comment by pohmelie** [comment]
@dpgeorge 
At first
``` diff
+    t->items[2] = MP_OBJ_NEW_SMALL_INT(0);
```
And this works fine after adding `-D_DIRENT_HAVE_D_INO=0` to makefile  :+1:

**Comment by dpgeorge** [comment]
Thanks @pohmelie!  Now merged in 64a909ef5113925adef19f275f62473de8ee68c5 with changes that (hopefully) fix the uos problem.  Please check that it does in fact work.

**Comment by pfalcon** [comment]
> !defined(_DIRENT_HAVE_D_INO) || _DIRENT_HAVE_D_INO

That's not native semantics of  _DIRENT_HAVE_D_INO. It's either defined or not. Overloading it like that is confusing. So, we either a) should follow how glibc uses it, or b) add MICROPY_DIRENT_HAVE_D_INO. I optimized for avoiding adding more adhoc config vars.

**Comment by dpgeorge** [comment]
> That's not native semantics of _DIRENT_HAVE_D_INO. It's either defined or not.

I understand, but this is a corner case (freedos is not posix, so it's trying to fit things together that don't work together).

Having it as `#ifdef _DIRENT_HAVE_D_INO` would mean defining it explicitly for all ports that do have d_ino.

**Comment by pfalcon** [comment]
> I understand, but this is a corner case (freedos is not posix, so it's trying to fit things together that don't work together).

But we should not handle D_TYPE in one way and D_INO in other way.

> Having it as #ifdef _DIRENT_HAVE_D_INO would mean defining it explicitly for all ports that do have d_ino.

Yes, just the same as _DIRENT_HAVE_D_TYPE.

**Comment by dpgeorge** [comment]
Ok, I'd rather do it as glibc does it than have our own MICROPY variable.  Easiest would be to define it in mpconfigport.h I guess.

**Comment by pfalcon** [comment]
Yes, again, just like _DIRENT_HAVE_D_TYPE is handled already. mpconfig_freedos.h just #undef's it.

**Comment by dpgeorge** [comment]
I can fix it if you didn't already.

**Comment by dpgeorge** [comment]
Fixed in 0d7de08e064dd124bd9a8c95bf6cef284a6bd4d9.

**Comment by pohmelie** [comment]
Excellent, have success with building out of ~~box~~ git clone :ok_hand:

**Comment by pfalcon** [comment]
@dpgeorge : Thanks!

**Comment by dpgeorge** [comment]
@pfalcon no problem, I should have waited for your reply before powering ahead :)
